### ISSUE TITLE
sql.Tx.QueryRow() results in panic

### ISSUE THREAD

**Issue by ahmedalhulaibi** [jackc/pgx#895]
Go: v1.15
Postgres: v10.12
___
I'm trying to upgrade from pgx v4.7.0 to 4.10.0 and I started experiencing a panic originating from [line 154](https://github.com/jackc/pgx/blob/e8f959e0e1a31026af1895c39b99c54f19fc2756/rows.go#L150):

https://github.com/jackc/pgx/blob/e8f959e0e1a31026af1895c39b99c54f19fc2756/rows.go#L150-L157

________
## Debug Report
The panic was specifically triggered when attempting to insert and violating a foreign key constraint. I have observed this with multiple queries that violate the foreign key constraint. I'm not sure if this is isolated to class 23 errors.

```
(dlv) print rows.err
error(*github.com/jackc/pgconn.PgError) *{
        Severity: "ERROR",
        Code: "23503",
        Message: "insert or update on table \"accounts\" violates foreign key constr...+24 more",
        Detail: "Key (user_id)=(404) is not present in table \"users\".",
        Hint: "",
        Position: 0,
        InternalPosition: 0,
        InternalQuery: "",
        Where: "",
        SchemaName: "public",
        TableName: "accounts",
        ColumnName: "",
        DataTypeName: "",
        ConstraintName: "accounts_users_fk",
        File: "ri_triggers.c",
        Line: 3266,
        Routine: "ri_ReportViolation",}
(dlv) 

```
____
It appears that `stmtcache` is `nil`. 
```
(dlv) print rows.conn.stmtcache
github.com/jackc/pgconn/stmtcache.Cache nil
```

My project doesn't use the statement cache, but I wouldn't expect to experience a panic if it is optional.

```go
// My projects initialization func
func InitializeDatabaseWithUrl(config *core.Configuration, url string) {
	var err error

	connConfig, err := pgx.ParseConfig(url)
	if err != nil {
		log.Fatal(err)
	}

	connConfig.BuildStatementCache = nil
	url = stdlib.RegisterConnConfig(connConfig)

	conn, err = sql.Open("pgx", url)
	if err != nil {
		log.Fatal(err)
	}

	conn.SetMaxOpenConns(25)
	conn.SetMaxIdleConns(25)
	conn.SetConnMaxLifetime(5 * time.Minute)
}
```
Below is a sample stack trace.
```
panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                    
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0xf889d4]                                                                                                                    
                                                                                                                                                                                           
goroutine 433 [running]:                                                                                                                                                                   
github.com/jackc/pgx/v4.(*connRows).Close(0xc000565c18)                                                                                                                                    
        /go/src/github.com/rainhq/rain/vendor/github.com/jackc/pgx/v4/rows.go:154 +0x214                                                                                                   
github.com/jackc/pgx/v4.(*connRows).Next(0xc000565c18, 0x1517ce0)                                                                                                                          
        /go/src/github.com/rainhq/rain/vendor/github.com/jackc/pgx/v4/rows.go:189 +0x198                                                                                                   
github.com/jackc/pgx/v4/stdlib.(*Conn).QueryContext(0xc000476120, 0x1517ce0, 0xc0000a8010, 0xc0002e0000, 0x1fb, 0xc00048e6c0, 0xe, 0xe, 0xc0004ef440, 0xc0003c1a50, ...)                   
        /go/src/github.com/rainhq/rain/vendor/github.com/jackc/pgx/v4/stdlib/sql.go:360 +0x37a                                                                                             
database/sql.ctxDriverQuery(0x1517ce0, 0xc0000a8010, 0x7f3f61355048, 0xc000476120, 0x0, 0x0, 0xc0002e0000, 0x1fb, 0xc00048e6c0, 0xe, ...)                                                  
        /usr/local/go/src/database/sql/ctxutil.go:48 +0x2f0                                                                                                                                
database/sql.(*DB).queryDC.func1()                                                                                                                                                         
        /usr/local/go/src/database/sql/sql.go:1643 +0x2c9                                                                                                                                  
database/sql.withLock(0x1508580, 0xc0004df5f0, 0xc00056d990)                                                                                                                               
        /usr/local/go/src/database/sql/sql.go:3284 +0x7f                                                                                                                                   
database/sql.(*DB).queryDC(0xc00011f790, 0x1517ce0, 0xc0000a8010, 0x1517ca0, 0xc0003fb100, 0xc0004df5f0, 0xc0003c1a60, 0xc0002e0000, 0x1fb, 0xc000205dd8, ...)                             
        /usr/local/go/src/database/sql/sql.go:1638 +0x834                                                                                                                                  
database/sql.(*Tx).QueryContext(0xc000104900, 0x1517ce0, 0xc0000a8010, 0xc0002e0000, 0x1fb, 0xc00056ddd8, 0xe, 0xe, 0x7f3f8af63560, 0xc0000a8ca0, ...)                                     
        /usr/local/go/src/database/sql/sql.go:2375 +0x1b7                                                                                                                                  
database/sql.(*Tx).QueryRowContext(...)                                                                                                                                                    
        /usr/local/go/src/database/sql/sql.go:2390                                                                                                                                         
database/sql.(*Tx).QueryRow(0xc000104900, 0xc0002e0000, 0x1fb, 0xc00056ddd8, 0xe, 0xe, 0x0)                                                                                                
        /usr/local/go/src/database/sql/sql.go:2401 +0xbf 
```
_____

## Recommended Solution

I'm not entirely familiar with the driver stack so my solution may lack the full context to offer a complete solution. I would naively patch this to validate that the `stmtcache` is not `nil` before proceeding with the `stmtcache` invalidation

```go
if rows.logger.shouldLog(LogLevelError) { 
    rows.logger.log(rows.ctx, LogLevelError, "Query", map[string]interface{}{"err": rows.err, "sql": rows.sql, "args": logQueryArgs(rows.args)}) 
} 
if rows.err != nil && rows.conn.stmtcache != nil { 
    rows.conn.stmtcache.StatementErrored(rows.sql, rows.err) 
} 
```

**Comment by jackc** [comment]
Good catch! Just released `v4.10.1` with the fix.

**Comment by ahmedalhulaibi** [comment]
Thanks for the quick fix!
